Reimplement stubs to improve performance

janvorli opened the pull request:
This change implements `FixupPrecodeStub`, `PrecodeStub`, `CallCountingStub` and VSD stubs `LookupStub`, `DispatchStub` and `ResolveStub` using a new mechanism with fixed code and separate RW data. The `LoaderHeap` was updated to support a new kind of allocation using interleaved code and data pages to support this new mechanism.
The JIT now generates code that uses indirection slot to jump to the methods using `FixupPrecode`, improving performance of the ASPNet plaintext benchmark by 3-4% depending on the target platform (measured on x64 Windows / Linux and arm64 Linux).

I have also removed the Holders, as the stubs are naturally properly aligned due to the way they are allocated.

There is now only a single variant of each stub, there are no long / short ones anymore as they are not needed - the indirect jumps we use now are not range limited.

Most of the stubs stuff is now target agnostic and the originally split implementation is now in single place for all targets. Only a few constants are defined as target specific in these.

The code for the stubs is no longer generated as bytes by C++ code, but rather written in asm and compiled. These precompiled templates are then used as a source to copy the code from. The x86 is a bit more complex than that due to the fact that it doesn't support PC relative indirect addressing, so we need to relocate all access to the data slots when generating the code pages.

As a further improvement, we could generate just a single page of the code and then just map it many times. This is left for future work.

ARM64 Unix differs from the other targets / platforms - there are various page sizes being used. So the asm templates are generated for 4k..64k page sizes and the variant is then picked at runtime based on the page size extracted from the OS.

This also removes a lot of writeable mappings created for modifications of the stub code when W^X is enabled, in the plaintext benchmark they were reduced by 75%. That results in a significant reducing of the .NET application startup time with W^X enabled.

I think the `LoaderHeap` would benefit from some refactoring, but I'd prefer leaving it for a follow up. It seems that for the sake of the review, it is better to keep it as is.

The change also implements logging of number of mappings and their exact locations. This helped me to drive the work and I am planning to use it for further changes. It can be removed in the future once we reach a final state.

There are still opportunities for improvement, but these stubs allowed me to scrape off the most significant portion of the mappings.

janvorli committed:
Reimplement stubs to improve performance

This change implements `FixupPrecodeStub`, `PrecodeStub`,
`CallCountingStub` and VSD stubs `LookupStub`, `DispatchStub` and
`ResolveStub` using a new mechanism with fixed code and separate RW
data. The `LoaderHeap` was updated to support a new kind of allocation
using interleaved code and data pages to support this new mechanism.
The JIT now generates code that uses indirection slot to jump to the
methods using `FixupPrecode`, improving performance of the ASPNet
plaintext benchmark by 3-4% depending on the target platform (measured
on x64 Windows / Linux and arm64 Linux).

I have also removed the Holders, as the stubs are naturally properly
aligned due to the way they are allocated.

There is now only a single variant of each stub, there are no long /
short ones anymore as they are not needed - the indirect jumps we use
now are not range limited.

Most of the stubs stuff is now target agnostic and the originally split
implementation is now in single place for all targets. Only a few
constants are defined as target specific in these.

The code for the stubs is no longer generated as bytes by C++ code, but
rather written in asm and compiled. These precompiled templates are then
used as a source to copy the code from. The x86 is a bit more complex
than that due to the fact that it doesn't support PC relative indirect
addressing, so we need to relocate all access to the data slots when
generating the code pages.

As a further improvement, we could generate just a single page of the
code and then just map it many times. This is left for future work.

ARM64 Unix differs from the other targets / platforms - there are
various page sizes being used. So the asm templates are generated for
4k..64k page sizes and the variant is then picked at runtime based on
the page size extracted from the OS.

This also removes a lot of writeable mappings created for modifications
of the stub code when W^X is enabled, in the plaintext benchmark they
were reduced by 75%. That results in a significant reducing of the .NET
application startup time with W^X enabled.

I think the `LoaderHeap` would benefit from some refactoring, but I'd
prefer leaving it for a follow up. It seems that for the sake of the
review, it is better to keep it as is.

The change also implements logging of number of mappings and their exact
locations. This helped me to drive the work and I am planning to use it
for further changes. It can be removed in the future once we reach a
final state.

There are still opportunities for improvement, but these stubs allowed
me to scrape off the most significant portion of the mappings.

janvorli committed:
Disable executable allocator statistics by default

janvorli commented:
Performance improvements - plaintext benchmark server side start time and client side read throughput (which is linearly proportional to requests/s). Please take the results with a grain of salt, they vary a lot and these are averages of 7 runs with the lowest outlier removed. But the trend is stable.

### Win x64
|          | Main |     | This PR | |
| ----- | ---- | ---- |----- | ---- | 
|  | W^X off | W^X on |  W^X off | W^X on |
| Start time [ms] | 324 | 386 | 324 | 357 |
| Read throughput [MB/s] | 662.27 | 640.35 | 674.22 | 679.10 |

### Linux Intel x64
|          | Main |     | This PR | |
| ----- | ---- | ---- |----- | ---- | 
|  | W^X off | W^X on |  W^X off | W^X on |
| Start time [ms] | 194 | 251 | 197 | 212 |
| Read throughput [MB/s] | 532.87 | 485.32 | 556.29 | 549.11 |

### Linux AMD x64
|          | Main |     | This PR | |
| ----- | ---- | ---- |----- | ---- | 
|  | W^X off | W^X on |  W^X off | W^X on |
| Start time [ms] | 143 | 182 | 143 | 155 |
| Read throughput [MB/s] | 933.55 | 882.35 | 960.85 | 921.60 |

janvorli commented:
I've measured similar trends on arm64 Linux in the past, but I need to re-measure the stuff after the recent cleanup and code unification changes.

janvorli commented:
It seems disabling the mapping loggings has broken the build, I am looking into it.

janvorli committed:
Fix build with executable allocator logging disabled

EgorBo commented on src/coreclr/vm/arm/thunktemplates.S:
```diff
@@ -0,0 +1,114 @@
+
+    LEAF_ENTRY StubPrecodeCode
+        ldr r12, DATA_SLOT(StubPrecode, MethodDesc)
+        ldr pc, DATA_SLOT(StubPrecode, Target)
```
I wonder if we can use a single `ldp` here to load both registers in a single instruction

janvorli replied:
We cannot, ldp requires consecutive register pair.

janvorli replied:
Ah, sorry, ldp doesn't even exist on arm32. You probably meant ldm, but that doesn't support PC relative addressing

janvorli committed:
Fix Windows ARM/ARM64 and macOS x64 builds

jkotas commented on src/coreclr/vm/amd64/thunktemplates.asm:
```diff
@@ -0,0 +1,80 @@
+        je     CountReachedZero
+        jmp    QWORD PTR [DATA_SLOT(CallCountingStub, TargetForMethod)]
+    CountReachedZero:
+        call   QWORD PTR [DATA_SLOT(CallCountingStub, TargetForThresholdReached)]
```
This has mismatched call/ret count. It is going to work particularly poorly with CET. (I understand that this problem was there even before your change.) 

Can the address of RemainingCallCountCell that is loaded into rax already be used as the stub identifying token instead?

janvorli replied:
That's a good point, it seems there already is a mean to get the stub from the call count cell address.

janvorli replied:
I made it work on x64, I'll implement it that way for all targets.

jkotas commented on src/coreclr/inc/jithelpers.h:
```diff
@@ -267,7 +267,7 @@
 
 #if defined(HAS_FIXUP_PRECODE)
-    JITHELPER(CORINFO_HELP_EE_PRECODE_FIXUP,    PrecodeFixupThunk,          CORINFO_HELP_SIG_NO_ALIGN_STUB)
+    JITHELPER(CORINFO_HELP_EE_PRECODE_FIXUP,    ThePreStub,                 CORINFO_HELP_SIG_NO_ALIGN_STUB)
```
All these `CORINFO_HELP_EE_...` entries can be deleted. They are left-over from fragile NGen.

jkotas commented on src/coreclr/utilcode/executableallocator.cpp:
```diff
@@ -499,9 +573,9 @@ void* ExecutableAllocator::ReserveWithinRange(size_t size, const void* loAddress
-        {
-            return NULL;
-        }
+	    {
```
Nit: The formatting looks off.

janvorli replied:
Somehow this file contained tabs at many places for some reason

jkotas commented on src/coreclr/vm/amd64/thunktemplates.asm:
```diff
@@ -0,0 +1,80 @@
+       jmp    QWORD PTR [DATA_SLOT(LookupStub, ResolveWorkerTarget)]
+LEAF_END_MARKED LookupStubCode, _TEXT
+
+LEAF_ENTRY DispatchStubCode, _TEXT
```
I am surprised that you were able to get away with the extra indirection in DispatchStubCode. It will be interesting to see the results of the microbenchmark runs.

janvorli replied:
I am not sure what indirection you mean. This was the original code:
https://github.com/dotnet/runtime/blob/a1a653c7217706fcb820914d65126234bc2447a8/src/coreclr/vm/amd64/virtualcallstubcpu.hpp#L286-L289
https://github.com/dotnet/runtime/blob/a1a653c7217706fcb820914d65126234bc2447a8/src/coreclr/vm/amd64/virtualcallstubcpu.hpp#L183-L191

jkotas replied:
I meant the performance overhead from the extra indirections introduced by this refactoring. The original code had two immediates, the new code has two indirections.  

@adamsitnik What's the best way to do dotnet/performance benchmark run for a PR like this one to see what improved/regressed on different target platforms?

adamsitnik replied:
> What's the best way to do dotnet/performance benchmark run for a PR like this one to see what improved/regressed on different target platforms?

We have two options:

1. Run the benchmarks using two different coreruns (before & after) and store them in dedicated folders, then compare them using ResultsComparer tool: https://github.com/dotnet/performance/blob/ef3edbc52d92c6b30ba1c316082f46865e5ff1d6/docs/benchmarking-workflow-dotnet-runtime.md#preventing-regressions It's the best solution if the implementation might change and you might need to re-run the benchmarks using updated corerun.
2. Run the benchmarks using two different coreruns without storing them in dedicated dolders, use BDN to perform the statistical test on the fly: https://github.com/dotnet/performance/blob/03207f183f042f6fc6b9f341df7a0e36b7175f5d/src/benchmarks/micro/README.md#private-runtime-builds it's good if you have only few benchmarks or don't intend to change the implementation

jkotas commented on src/coreclr/vm/loaderallocator.cpp:
```diff
@@ -1118,7 +1118,8 @@ void LoaderAllocator::Init(BaseDomain *pDomain, BYTE *pExecutableHeapMemory)
     // Make sure that we reserve as little as possible on 32-bit to save address space
-    _ASSERTE(dwTotalReserveMemSize <= VIRTUAL_ALLOC_RESERVE_GRANULARITY);
+    // We cannot reserve less than needed
+    //_ASSERTE(dwTotalReserveMemSize <= VIRTUAL_ALLOC_RESERVE_GRANULARITY);
```
Delete this instead of uncommenting it?

janvorli replied:
This is an accidental leftover, thank you for noticing.

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -8804,7 +8804,15 @@ void CEEInfo::getFunctionEntryPoint(CORINFO_METHOD_HANDLE  ftnHnd,
     ftn = ftn->GetMethodTable()->MapMethodDeclToMethodImpl(ftn);
 
-    ret = (void *)ftn->TryGetMultiCallableAddrOfCode(accessFlags);
+    if (!ftn->IsFCall() && ftn->MayHavePrecode() && ftn->GetPrecodeType() == PRECODE_FIXUP)
```
I think this condition is not right for the tiered compilation disabled case (not default, but still used by some first parties). Does this need to call into code version manager?

janvorli replied:
It is possible (although all the tests pass with tiered compilation disabled except for one where the failure looks unrelated to this). What do you think is the specific problem here?

jkotas replied:
I think that the tiered compilation disabled case will go through unnecessary indirection when the target method is JITed already.

janvorli replied:
I've tried to add `!ftn->HasStableEntryPoint()` to the condition. It severely degraded the plaintext benchmark performance. I've also thought that `!ftn->IsPointingToStableNativeCode()` might be needed, but the function seems to never report true together with the preexisting condition (based on my testing when I've added an assert and none of the coreclr pri1 tests hit it, with or without tiered compilation enabled).

jkotas replied:
I think the logic should be something like this:
```
if (ftn->IsVersionable())
{
    IAT_PVALUE // must go via indirection to enable versioning 
}
else
{
    if (the target has final code)
        IAT_VALUE for the final code // this includes FCalls and method that are already JITed (w/o tiering)
    else
        IAT_PVALUE // must go via indirection to pick up the final code once it is ready
}
```

janvorli replied:
I will give it a try

janvorli committed:
Reflect PR feedback and few fixes

* Change the CallCountingStub to not to use return address of an
unbalanced call as a stub identifying token. The counter address is used
instead on all targets.
* Fix some tabs instead of spaces
* Fix getTargetMethodDesc - in some cases, we get address of the start
of the FixupPrecode too.
* Remove a leftover comment

EgorBo commented:
@janvorli since you seem to be refactoring a lot of stuff is it possible now to get an understanding where a currently jitted method will end up in memory precisely (or at lest approximately)? 
Basically, exposed AllocPtr of a LoaderCodeHeap 

to be able to rely on it for "relocs" in jitted code

janvorli commented:
@EgorBo this change only changes allocation of stubs, it doesn't change in any way where jitted code will end up. But I am not sure if I fully understood what you are after.

janvorli committed:
Fix macOS x64

The assembler was generating 32 bit conditional relative jumps instead
of ones with 8 bit displacement. I've found that a presence of a global
label between the jump site and the destination makes the assembler to
do that. Changing PATCH_LABEL macro fixed it.

janvorli committed:
Fix ARM64 StubPrecodeCode_End extraction

janvorli committed:
Reflect feedback and improve some JIT helpers perf

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -11126,6 +11133,23 @@ void* CEEJitInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,         /* IN  */
+        {
+            Precode* pPrecode = Precode::GetPrecodeFromEntryPoint((PCODE)hlpDynamicFuncTable[dynamicFtnNum].pfnHelper);
+            _ASSERTE(pPrecode->GetType() == PRECODE_FIXUP);
+            *ppIndirection = ((FixupPrecode*)pPrecode)->GetTargetSlot();
```
What guarantees that we have the final code of the method by this point?

If this is a valid optimization, we should be doing it when filling `hlpDynamicFuncTable`.

jkotas replied:
> This just passes the fixup precode indirection slot to the JIT. If we don't have the final code, the slot points to the fixup part of the slot.

Ah ok, I have missed that.

janvorli committed:
Revert VSD stubs changes

Extensive benchmarking has shown that the new model for VSD stubs is
causing steady state performance degradation for code using a lot of
virtual calls. It wasn't showing up in the plaintext benchmark I have
used as the driver of the change.
This change looses part of the startup perf improvements, but I am
planning to add back the Lookup stubs in a follow up change. Those
should not be perf critical and they are likely the ones where the
startup improvements are gained from.

janvorli commented:
I have removed the VSD stubs change, it was causing steady state performance degradation in more complex ASPNet benchmarks like the Orchard or Fortunes. I am planning to re-add the Lookup stub in a follow up change, but for this PR, it was safer to remove it all.

jkotas commented on src/coreclr/vm/common.h:
```diff
@@ -417,6 +417,30 @@ extern DummyGlobalContract ___contract;
 
 #endif // defined(_DEBUG)
 
+#define ENUM_PAGE_SIZES \
+    ENUM_PAGE_SIZE(4096)  \
+    ENUM_PAGE_SIZE(8192)  \
+    ENUM_PAGE_SIZE(16384) \
+    ENUM_PAGE_SIZE(32768) \
+    ENUM_PAGE_SIZE(65536)
+
+inline void FillStubCodePage(BYTE* pageBase, const void* code, int codeSize, int pageSize)
+{
+    int totalCodeSize = (pageSize / codeSize) * codeSize;
+
+    memcpy(pageBase, code, codeSize);
+
+    int i;
+    for (i = codeSize; i < pageSize / 2; i *= 2)
+    {
+        memcpy(pageBase + i, pageBase, i);
+    }
+
+    if (i != totalCodeSize)
+    {
+        memcpy(pageBase + i, pageBase, totalCodeSize - i);
+    }
+}
 
```
Is there a better place for this than vm/common.h?

janvorli replied:
Good point, this was just a quick temporary location when I was doing the final unification of some code and then I've forgotten about it.

janvorli committed:
Move FillStubCodePage to a better location

janvorli committed:
Merge branch 'main' into new-stubs

EgorBo commented:
![image](https://user-images.githubusercontent.com/523221/159427502-b9a36aac-2c10-4b3b-96a2-be84e747178b.png)

I assume this perf improvement is because of this PR 👍 (according to the graph, the commit range is https://github.com/dotnet/runtime/compare/05cb7f5...110cb9f)

kunalspathak commented:
windows x64 improvements: https://github.com/dotnet/perf-autofiling-issues/issues/4226

DrewScoggins commented:
Windows Arm64 Improvements https://github.com/dotnet/perf-autofiling-issues/issues/4251

DrewScoggins commented:
Ubuntu Arm64 improvements https://github.com/dotnet/perf-autofiling-issues/issues/4258

janvorli commented:
@Jozkee I am sorry for missing your message before. Yes, I believe the regression was caused by that change. There are few corner cases where the new stubs implementation has slight negative effect on micro-benchmarks. Some delegate calls are one of those cases and this test was one of those where I've seen it to have visible impact when I was developing the change.
